keps/0019-package-api-versioning.md

```diff
@@ -2,64 +2,374 @@
+```
+
+Installs `kafka-3.0.0-1.1.0`.
+
+**Install a Kafka operator with a specific version**
+
+Let's say a user wants to use Apache Kafka `3.0.0` and even though there's a `3.0.0` with a `1.1.0` revision released, they want the operator with revision `1.0.0` due to reasons like:
+- revision `1.1.0` introduced a bug for which there's still no released fixes
+- revision `1.1.0` changed in a way that will require them to invest time and resources to adapt
+
+```
+kudo install kafka --version 3.0.0-1.0.0
+```
+
+**Install a Kafka operator providing a partial version**
+
+Another interesting thing that having the concept of _version_ contain the _app version_ concatenated with the _operator revision_ is that it would be possible to not have operator users necessarily have to know about the existence of _app version_ and _operator revision_. With regards to versioning, CLI interactions could optionally "flatten" the usage of both flags into just the `--version` flag, which could also be provided as a prefix to be matched.
+
+```
+kudo install kafka --version 3
+```
+
+Installs the latest Kafka operator based on Apache Kafka 3.x.x.
+
+```
+kudo install kafka --version 2.3
+```
+
+Installs the latest Kafka operator based on Apache Kafka 2.3.x.
+
+```
+kudo install kafka --version 2.3.1
+```
+
+Installs the latest Kafka operator based on Apache Kafka 2.3.1.
 
-Through the `apiVersion` field in an `operator.yaml` it is already guaranteed that the operator will match a specific API version of KUDO. However, this field isn't taken into account when installing packages from a repository. KUDO will try to install the latest version of the package even if the `apiVersion` doesn't match. This can be mitigated by adding the packages `apiVersion` to the repository index and filtering out all packages that don't have a matching version. This way, older versions of packages can still be part of the repository and keep working with older versions of KUDO.
+**Upgrade Kafka operator**
+
+The `kudo upgrade` commands would look similar as the `kudo install` commands above.
+
+**Search package registry**
+
+```
+$ kudo search kafka
+kafka-2.3.0-1.0.0
+kafka-2.3.0-1.1.0
+kafka-2.3.0-1.1.1
+kafka-2.3.1-1.2.0
+kafka-3.0.0-1.0.0
+kafka-3.0.0-1.1.0
+kafka-3.0.1-1.1.0
+kafka-3.1.1-1.0.0
+```
+
+```
+$ kudo search kafka --version 3
+kafka-3.0.0-1.0.0
+kafka-3.0.0-1.1.0
+kafka-3.0.1-1.1.0
+kafka-3.1.1-1.0.0
+```
+
+```
+$ kudo search kafka --version 3.1
+kafka-3.1.1-1.0.0
+```
+
+```
+$ kudo search kafka --version 2.3
+kafka-2.3.0-1.0.0
+kafka-2.3.0-1.1.0
+kafka-2.3.0-1.1.1
+kafka-2.3.1-1.2.0
+```
+
+---
+
+An alternative notation for specifying the version could also look something like this, but this is not important:
+
+```
+kudo install kafka@3
+```
+```
+kudo install kafka@2.3.0
+```
+```
+kudo install kafka@2.3.1
+```
+
+And, orthogonal to the singular VS composite SemVer proposals:
+
+### Support multiple KUDO API versions in the package repository
```

Comment:
yep, I think the solution here is to add it to index and provide filtering on CLI 👍 I would remove this sentence `KUDO will try to install the latest version of the package even if the `apiVersion` doesn't match.` because it feels like that is the target behavior. We should rather outline that we plan to filter on that